airflow/models.py

```diff
@@ -1652,13 +1645,48 @@ def resolve_template_files(self):
 
     @property
     def upstream_list(self):
-        """@property: list of tasks directly upstream"""
-        return self._upstream_list
+        """@property: list of tasks of the same schedule directly upstream"""
+        tasks = [task for (task, _) in self._upstream_tuples]
+        return list(set(tasks))
 
     @property
     def downstream_list(self):
-        """@property: list of tasks directly downstream"""
-        return self._downstream_list
+        """@property: list of tasks of the same schedule directly downstream"""
+        tasks = [task for (task, _) in self._downstream_tuples]
+        return list(set(tasks))
+
+    @property
+    def upstream_triggers(self):
+        """@property: all the triggers on which this task depends on that
+        were added through set_downstream and set_upstream only"""
+        return [trigger for (_, trigger) in self.upstream_tuples]
+
+    @property
+    def upstream_triggers_extra(self):
+        """@property: all the triggers on which this task depends on that
+        were added through depends_on_past or wait_for_downstream only"""
+        return [trigger for (_, trigger) in self.upstream_tuples_extra]
+
+    @property
+    def upstream_tuples(self):
+        """@property: tuples of all the (task, trigger) on which this task
+        depends on that were added through set_downstream and set_upstream
+        only"""
+        return self._upstream_tuples
+
+    @property
+    def upstream_tuples_extra(self):
+        """@property: tuples of all the (task, trigger) on which this task
+        depends on that were added through depends_on_past or
+        wait_for_downstream only"""
+        tuples = self._upstream_tuples_extra
```

Comment:
maybe we could move the 
````        if self.depends_on_past:
            trigger = Trigger(self, state=State.SUCCESS, past_executions=1)
            self._upstream_tuples_extra.append((self, trigger))
            self._downstream_tuples_extra.append((self, trigger))
````
here such that we would no longer need the `self._upstream_tuples_extra` but directly build the `tuples` list on the fly here ?